[Done] Create polygons from indices

caspervdw opened the pull request:
This build onto #322

brendan-ward commented:
@caspervdw note: I think there is a known issue in GEOS (at least on IRC) with ambiguity around ownership transfer of inputs in cases that raise exceptions; this came up in discussions around #306.  I think this might impact your fix in c11b1c8: in the short term there will be memory leaks because of this, but in the long term handling those ourselves will cause segfaults.  Sorry that I haven't had a chance to log a more complete error report to GEOS; wanted to itemize the different cases...

caspervdw commented:
With https://github.com/pygeos/pygeos/commit/c11b1c8fe28d27bfae35b2c4d101ab7cbb6c5a2d, there are no segfaults taking place, so I conclude that cleaning up after a failed Polygon construction is necessary. Still, even with this cleanup, I observe memory leaks coming from the polygon constructor. So I think there is something intermediate that is not cleaned up properly by GEOS.

So I also added a pre-emptive check on the input type, so that it can't fail. Now we still have one left:

```
___________________ test_polygons[holes0-indices0-expected0] ___________________
[VALGRIND LEAK]

Valgrind has detected a memory leak:

**1** 
**1** **********************************************************************
**1** pygeos/test/test_creation_indices.py::test_polygons[holes0-indices0-expected0]
**1** **********************************************************************
==1== 65 (+65) (32 (+32) direct, 33 (+33) indirect) bytes in 1 (+1) blocks are definitely lost in loss record 6,856 of 17,475
==1==    at 0x4C2BBAF: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0xF40E3DD: _buffer_get_info (in /usr/local/lib/python3.6/site-packages/numpy/core/_multiarray_umath.cpython-36m-x86_64-linux-gnu.so)
==1==    by 0xF40EB74: array_getbuffer (in /usr/local/lib/python3.6/site-packages/numpy/core/_multiarray_umath.cpython-36m-x86_64-linux-gnu.so)
==1==    by 0x37D890FE: __pyx_memoryview___pyx_pf_15View_dot_MemoryView_10memoryview___cinit__ (_geometry.c:11849)
==1==    by 0x37D890FE: __pyx_memoryview___cinit__ (_geometry.c:11781)
==1==    by 0x37D890FE: __pyx_tp_new_memoryview (_geometry.c:22650)
==1==    by 0x4F7BB84: type_call (typeobject.c:895)
==1==    by 0x37D7F519: __Pyx_PyObject_Call (_geometry.c:24783)
==1==    by 0x37D7F519: __pyx_memoryview_new (_geometry.c:15923)
==1==    by 0x37D7F765: __Pyx_ValidateAndInit_memviewslice.constprop.75 (_geometry.c:28115)
==1==    by 0x37D7FBBA: __Pyx_PyObject_to_MemoryviewSlice_ds_object.constprop.74 (_geometry.c:28178)
==1==    by 0x37D90054: __pyx_pf_6pygeos_9_geometry_6polygons_1d.isra.44 (_geometry.c:7478)
==1==    by 0x37D929EA: __pyx_pw_6pygeos_9_geometry_7polygons_1d (_geometry.c:6887)
==1==    by 0x4F6C9FB: _PyCFunction_FastCallDict (methodobject.c:231)
==1==    by 0x4FB12D9: call_function (ceval.c:4851)
==1== 
==1== 65 (+65) (32 (+32) direct, 33 (+33) indirect) bytes in 1 (+1) blocks are definitely lost in loss record 6,857 of 17,475
==1==    at 0x4C2BBAF: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1==    by 0xF40E3DD: _buffer_get_info (in /usr/local/lib/python3.6/site-packages/numpy/core/_multiarray_umath.cpython-36m-x86_64-linux-gnu.so)
==1==    by 0xF40EB74: array_getbuffer (in /usr/local/lib/python3.6/site-packages/numpy/core/_multiarray_umath.cpython-36m-x86_64-linux-gnu.so)
==1==    by 0x37D890FE: __pyx_memoryview___pyx_pf_15View_dot_MemoryView_10memoryview___cinit__ (_geometry.c:11849)
==1==    by 0x37D890FE: __pyx_memoryview___cinit__ (_geometry.c:11781)
==1==    by 0x37D890FE: __pyx_tp_new_memoryview (_geometry.c:22650)
==1==    by 0x4F7BB84: type_call (typeobject.c:895)
==1==    by 0x37D7F519: __Pyx_PyObject_Call (_geometry.c:24783)
==1==    by 0x37D7F519: __pyx_memoryview_new (_geometry.c:15923)
==1==    by 0x37D7F765: __Pyx_ValidateAndInit_memviewslice.constprop.75 (_geometry.c:28115)
==1==    by 0x37D7FBBA: __Pyx_PyObject_to_MemoryviewSlice_ds_object.constprop.74 (_geometry.c:28178)
==1==    by 0x37D900A8: __pyx_pf_6pygeos_9_geometry_6polygons_1d.isra.44 (_geometry.c:7490)
==1==    by 0x37D929EA: __pyx_pw_6pygeos_9_geometry_7polygons_1d (_geometry.c:6887)
==1==    by 0x4F6C9FB: _PyCFunction_FastCallDict (methodobject.c:231)
==1==    by 0x4FB12D9: call_function (ceval.c:4851)
==1== 
==1== LEAK SUMMARY:
==1==    definitely lost: 2,936 (+64) bytes in 30 (+2) blocks
==1==    indirectly lost: 1,579 (+66) bytes in 20 (+4) blocks
==1==      possibly lost: 6,933,057 (-1,740) bytes in 67,170 (-1) blocks
==1==    still reachable: 16,390,388 (+7,563) bytes in 114,529 (+67) blocks
==1==                       of which reachable via heuristic:
==1==                         newarray           : 853 (+0) bytes in 40 (+0) blocks
==1==                         multipleinheritance: 2,632 (+0) bytes in 45 (+0) blocks
==1==         suppressed: 0 (+0) bytes in 0 (+0) blocks
==1== Reachable blocks (those to which a pointer was found) are not shown.
==1== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==1== 
**1** 
**1** **********************************************************************
=========================== short test summary info ============================
FAILED pygeos/test/test_creation_indices.py::test_polygons[holes0-indices0-expected0]
============== 1 failed, 13 passed, 7 xfailed in 65.79s (0:01:05) ==============
```

brendan-ward commented:
Sorry I wasn't very clear.  See https://trac.osgeo.org/geos/ticket/1111.

What I meant was that once that issue is fixed, freeing passed-in geometries ourselves will segfault.  Right now they don't, and they lead to a memory leak if the call to GEOS returns `NULL`.

If you'd like, you can version-guard this against GEOS, and cleanup ourselves for versions known to leak.  Note that there may be an intermediate GEOS version where this is leaking, but older ones may be OK (not sure: one would need to test or check the GEOS implementation at that version).

brendan-ward commented:
I'm not seeing the memory leak on Python 3.9 / GEOS 3.7  (`Dockerfile.valgrind` bumped to `python:3.9-slim-buster` base image):
```
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /code
plugins: valgrind-0.1.0
collecting ... collected 8 items

pygeos/test/test_creation_indices.py::test_polygons[holes0-indices0-expected0] PASSED [ 12%]
pygeos/test/test_creation_indices.py::test_polygons[holes1-indices1-expected1] PASSED [ 25%]
pygeos/test/test_creation_indices.py::test_polygons[holes2-indices2-expected2] PASSED [ 37%]
pygeos/test/test_creation_indices.py::test_polygons[holes3-indices3-expected3] PASSED [ 50%]
pygeos/test/test_creation_indices.py::test_polygons[holes4-indices4-expected4] PASSED [ 62%]
pygeos/test/test_creation_indices.py::test_polygons[holes5-indices5-expected5] PASSED [ 75%]
pygeos/test/test_creation_indices.py::test_polygons[holes6-indices6-expected6] PASSED [ 87%]
pygeos/test/test_creation_indices.py::test_polygons[holes7-indices7-expected7] PASSED [100%]

============================== 8 passed in 20.10s ==============================
```

Also not seeing leaks on Python 3.6 / GEOS 3.5 (current `Dockerfile.valgrind`)

```
============================= test session starts ==============================
platform linux -- Python 3.6.12, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /code
plugins: valgrind-0.1.0
collecting ... collected 8 items

pygeos/test/test_creation_indices.py::test_polygons[holes0-indices0-expected0] PASSED [ 12%]
pygeos/test/test_creation_indices.py::test_polygons[holes1-indices1-expected1] PASSED [ 25%]
pygeos/test/test_creation_indices.py::test_polygons[holes2-indices2-expected2] PASSED [ 37%]
pygeos/test/test_creation_indices.py::test_polygons[holes3-indices3-expected3] PASSED [ 50%]
pygeos/test/test_creation_indices.py::test_polygons[holes4-indices4-expected4] PASSED [ 62%]
pygeos/test/test_creation_indices.py::test_polygons[holes5-indices5-expected5] PASSED [ 75%]
pygeos/test/test_creation_indices.py::test_polygons[holes6-indices6-expected6] PASSED [ 87%]
pygeos/test/test_creation_indices.py::test_polygons[holes7-indices7-expected7] PASSED [100%]

============================== 8 passed in 28.61s ==============================
```

Makes it a bit challenging to help identify the source of the leak...  I'm assuming you are using the Python 3.6 / GEOS 3.5 image?

brendan-ward commented on pygeos/test/test_creation_indices.py:
```diff
@@ -117,6 +115,50 @@ def test_linearrings_invalid(coordinates):
+    actual = pygeos.polygons([linear_ring, linear_ring], holes, indices=indices)
+    assert_geometries_equal(actual, expected)
+
+
```
It would be good to add a test for missing shells too.  E.g.,

```python
def test_missing_shell():
    actual = pygeos.polygons([None, linear_ring], [hole_1, hole_2], indices=[0, 1])
    assert_geometries_equal(actual, [None, poly_hole_2])
```

This currently fails.

brendan-ward commented on pygeos/_geometry.pyx:
```diff
@@ -321,3 +323,120 @@ def collections_1d(object geometries, object indices, int geometry_type = 7):
+            # return None for missing shells (ignore possibly present holes)
+            if shell == NULL:
+                result_view[poly_idx] = PyGEOS_CreateGeometry(NULL, geos_handle)
+                continue
```
```suggestion
                hole_idx_1 += hole_count[poly_idx]
                continue
```

See failing test below

caspervdw replied:
Good catch! I corrected it and added the test.

brendan-ward reviewed (changes requested):
Looks like there is a counter that needs to be incremented for empty shells, but otherwise this looks good (subject to other comments about GEOS-related leaks).

caspervdw commented:
@brendan-ward Your remark about cleaning up after a GEOS Exception got me started on the original polygon creators. 

A generic solution to this is preventing GEOS Exceptions to occur in the first place. This we can do by pre-emptively check the type of the input. We already did this for the collections, but for the polygons we did not. I added this check in the original polygons `ufunc`. Because this was not possible for the `polygons_without_holes`, I removed that function and routed everything through the `polygons` (formerly: `polygons_with_holes`). While I was at it, I noticed that `polygons(None, [None])` segfaulted. I fixed that too. Also, I removed the GIL in `polygons` and `create_collection`.

Sorry for the big heap of changes, it was all so alike that I just went with the flow... I hope you can bear with me on this.

brendan-ward commented on src/ufuncs.c:
```diff
@@ -2261,30 +2311,49 @@ static void create_collection_func(char** args, npy_intp* dimensions, npy_intp*
-    OUTPUT_Y;
+    geom_arr[i] = GEOSGeom_createCollection_r(ctx, type, temp_geoms, n_geoms);
+    if (geom_arr[i] == NULL) {
+      // We will have a memory leak now (https://trac.osgeo.org/geos/ticket/1111)
```
```suggestion
      // We may have a memory leak now (https://trac.osgeo.org/geos/ticket/1111)
```

(We know polygons are problematic; not sure about collections)

brendan-ward reviewed (approved):
Thanks @caspervdw for the improvements.  Great idea on pre-emptively checking types prior to calling GEOS, to sidestep the bug there.

Thanks also for taking the chance to refactor the poly ufuncs and release the GIL.

jorisvandenbossche commented:
In contrast to the other creation functions, when using `indices`, with `polygons(..)` your input needs to be split between shells and holes (exterior and interior rings). 

Personally I find this a bit strange (or inconvenient, for the use case I was trying it out with). 

Using the example from https://github.com/pygeos/pygeos/issues/343, if starting with a single 1D array of linearrings and indices:

```python
import pygeos

# 1D array of rings with shape (10, )
rings = pygeos.linearrings(np.random.randn(10, 3, 2))
polygon_indices = np.array([0,1,1,2,2,2,3,4,4,5])
```

with the current API, I need to do:

```python
polygon_indices_shell = np.insert(np.diff(polygon_indices), 0, 1)
shells_mask = polygon_indices == 1
holes_mask = polygon_indices == 0

shells = rings[shells_mask]
holes = rings[holes_mask]
hole_indices = polygon_indices[holes_mask]

result =  pygeos.polygons(list(shells), list(holes), indices=hole_indices)
```

but alternatively (and what I initially expected to pass), it could also be an option to pass the shells/holes as a single array, so it would look like (for this starting data):

```
pygeos.polygons(rings, indices=polygon_indices)
```

(one constraint with this is that the shell always needs to come before the holes in the array)

Of course, it depends on your actual use case (and what your starting data is) which API would be most convenient. So there might be good use cases for both ways.

One thing is that the alternative with a single array of shells/holes combined would be more similar with eg the reverse of `get_parts` (or if a `get_rings` would be added as in https://github.com/pygeos/pygeos/pull/342, this would provide an exact reverse operation).

caspervdw commented:
What you propose is the same as I originally wrote in #241. 
So I agree with you on this, the API that you describe (with a single array of rings) is cleaner.

In a way, the polygons are 'just' collections of linearrings. We should treat them as such in the `collections_1d` and `get_parts` functions.
